File: go/schema.go
package audit

import "regexp"

// AuditActorType identifies who performed an action.
type AuditActorType string

const (
	ActorUser        AuditActorType = "user"
	ActorAgent       AuditActorType = "agent"
	ActorSystem      AuditActorType = "system"
	ActorIntegration AuditActorType = "integration"
	ActorAPIClient   AuditActorType = "api_client"
)

// AuditOutcome is the result of an action.
type AuditOutcome string

const (
	OutcomeSuccess AuditOutcome = "success"
	OutcomeFailure AuditOutcome = "failure"
	OutcomeDenied  AuditOutcome = "denied"
)

// AuditResource is the thing an action was performed against. The Type is a
// namespaced kind (e.g. "crm.contact"); the ID is its canonical identifier.
type AuditResource struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// AuditDiff is a structural diff captured at write time. Either side may be
// absent (omitted on create/delete); a *present* null (e.g. After on a delete)
// is meaningful and is serialized as JSON null, never dropped.
type AuditDiff struct {
	Before any `json:"before,omitempty"`
	After  any `json:"after,omitempty"`
}

// AuditEvent is the canonical audit event — the shared shape every language SDK
// serializes. It carries ZERO customer content by design: only identity,
// resource references, outcome, and namespaced action + metadata the emitter
// supplies. Field names and json tags are part of the cross-language hash
// contract and are verified byte-for-byte against ../spec/parity-corpus.json.
//
// Optional fields use omitempty so an absent field is dropped before
// canonicalization (matching the TS "undefined → omitted" rule). The event fed
// to ComputeEventHash is this struct with HashCurrent empty (thus omitted).
type AuditEvent struct {
	ID             string         `json:"id"`
	OrganizationID string         `json:"organizationId"`
	ActorType      AuditActorType `json:"actorType"`
	ActorID        string         `json:"actorId"`
	ActorEmail     string         `json:"actorEmail,omitempty"`
	Action         string         `json:"action"`
	Resource       AuditResource  `json:"resource"`
	Outcome        AuditOutcome   `json:"outcome"`
	Reason         string         `json:"reason,omitempty"`
	SessionID      string         `json:"sessionId,omitempty"`
	ConversationID string         `json:"conversationId,omitempty"`
	IPAddress      string         `json:"ipAddress,omitempty"`
	UserAgent      string         `json:"userAgent,omitempty"`
	GeoCountry     string         `json:"geoCountry,omitempty"`
	Diff           *AuditDiff     `json:"diff,omitempty"`
	Metadata       map[string]any `json:"metadata"`
	Timestamp      string         `json:"timestamp"`
	// HashPrevious links this event to the prior one in its per-org-per-day
	// chain. Absent (nil) on the first event of a chain — not null.
	HashPrevious *string `json:"hashPrevious,omitempty"`
	// HashCurrent is SHA-256 of canonical-JSON(this event minus HashCurrent).
	// Empty (thus omitted) while computing the hash; set afterward.
	HashCurrent string `json:"hashCurrent,omitempty"`
}

// Baseline event actions — the generic surface every app shares. Emitters are
// NOT limited to these: any consumer defines its own namespaced actions
// ("billing.invoice_voided", "fieldops.task_submitted", …) and emits them
// directly; canonicalization treats Action as an opaque string. These baseline
// names stay stable because dashboards/alerts/reports pivot off them.
const (
	ActionUserSignin           = "user.signin"
	ActionUserSignout          = "user.signout"
	ActionUserPasswordChanged  = "user.password_changed"
	ActionUserInvited          = "user.invited"
	ActionOrgCreated           = "org.created"
	ActionOrgMemberAdded       = "org.member_added"
	ActionOrgMemberRemoved     = "org.member_removed"
	ActionOrgRoleChanged       = "org.role_changed"
	ActionOrgSubscriptionChgd  = "org.subscription_changed"
	ActionOrgProductPurchased  = "org.product_purchased"
	ActionAgentConfigChanged   = "agent.config_changed"
	ActionAgentKnowledgeAdded  = "agent.knowledge_doc_added"
	ActionAgentKnowledgeRemvd  = "agent.knowledge_doc_removed"
	ActionAgentEscalationMade  = "agent.escalation_created"
	ActionAgentToolFailed      = "agent.tool_failed"
	ActionCRMContactCreated    = "crm.contact_created"
	ActionCRMContactMerged     = "crm.contact_merged"
	ActionCRMContactDeleted    = "crm.contact_deleted"
	ActionAPIKeyMinted         = "api.key_minted"
	ActionAPIKeyRotated        = "api.key_rotated"
	ActionAPIKeyRevoked        = "api.key_revoked"
	ActionIntegrationConnected = "integration.connected"
	ActionIntegrationDisconn   = "integration.disconnected"
)

var namespacedActionRe = regexp.MustCompile(`^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]+)+$`)

// IsNamespacedAction validates the "namespace.verb" action convention (a
// lowercase namespace and at least one lowercase verb segment, dot-separated,
// e.g. "crm.contact_created", "google.gmail.message_sent"). Assert this at your
// trust boundary; canonicalization itself treats Action as an opaque string, so
// this is a convention check, not a hard schema constraint.
func IsNamespacedAction(action string) bool {
	return namespacedActionRe.MatchString(action)
}

File: src/schema.ts
import { z } from "zod";

/**
 * Smoo Audit — canonical event schema.
 *
 * The shared shape every language SDK emits and every consumer queries. This
 * TypeScript schema is the source of truth; the Python / Rust / Go / .NET ports
 * mirror it field-for-field and are verified byte-for-byte against the shared
 * parity corpus (`spec/parity-corpus.json`).
 *
 * The schema is GENERIC — zero customer/vertical content. `action` is an opaque
 * `namespace.verb` string (see {@link isNamespacedAction}); {@link AUDIT_ACTIONS}
 * is only a baseline of the actions every app shares, not a closed set.
 */

/** Who performed the action. */
export const auditActorTypeSchema = z.enum([
  "user",
  "agent",
  "system",
  "integration",
  "api_client",
]);
export type AuditActorType = z.infer<typeof auditActorTypeSchema>;

/** Outcome of the action. */
export const auditOutcomeSchema = z.enum(["success", "failure", "denied"]);
export type AuditOutcome = z.infer<typeof auditOutcomeSchema>;

/**
 * Resource the action was performed against (e.g. a CRM contact, an agent
 * config). `id` is the canonical identifier in the relevant table.
 */
export const auditResourceSchema = z.object({
  type: z.string(),
  id: z.string(),
});
export type AuditResource = z.infer<typeof auditResourceSchema>;

/**
 * Structural diff captured at write time. Either side may be omitted (on create
 * `before` is absent; on delete `after` is absent). Stored as opaque JSON.
 */
export const auditDiffSchema = z.object({
  before: z.unknown().optional(),
  after: z.unknown().optional(),
});
export type AuditDiff = z.infer<typeof auditDiffSchema>;

/**
 * The wire payload an emitter sends. The server enriches actor / request / hash
 * fields from the authenticated request context — clients NEVER set those.
 */
export const auditEventInputSchema = z.object({
  /** `namespace.verb` action string, e.g. `crm.contact_created`. Opaque to the serializer. */
  action: z.string(),
  resource: auditResourceSchema,
  /** Default 'success' if omitted. */
  outcome: auditOutcomeSchema.optional(),
  /** Human-readable reason (especially for 'failure' / 'denied'). */
  reason: z.string().optional(),
  sessionId: z.string().optional(),
  conversationId: z.string().optional(),
  diff: auditDiffSchema.optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
});
export type AuditEventInput = z.infer<typeof auditEventInputSchema>;

/**
 * The full event shape as persisted / hashed. All identity + request fields are
 * present after server-side enrichment. `hashCurrent` is the SHA-256 of
 * canonical-JSON(this event minus `hashCurrent`); it is excluded from the hash
 * input and therefore optional on the pre-hash object fed to the serializer.
 */
export const auditEventSchema = z.object({
  /** ULID-like sortable identifier. */
  id: z.string(),
  organizationId: z.string(),
  actorType: auditActorTypeSchema,
  actorId: z.string(),
  actorEmail: z.string().optional(),
  action: z.string(),
  resource: auditResourceSchema,
  outcome: auditOutcomeSchema,
  reason: z.string().optional(),
  sessionId: z.string().optional(),
  conversationId: z.string().optional(),
  ipAddress: z.string().optional(),
  userAgent: z.string().optional(),
  geoCountry: z.string().optional(),
  diff: auditDiffSchema.optional(),
  metadata: z.record(z.string(), z.unknown()),
  /** ISO 8601 UTC timestamp. */
  timestamp: z.string(),
  /** Previous event's hash in the per-org-per-day chain. Absent on the first event. */
  hashPrevious: z.string().optional(),
  /** SHA-256 of canonical-JSON(this event minus hashCurrent). Absent pre-seal. */
  hashCurrent: z.string().optional(),
});
export type AuditEvent = z.infer<typeof auditEventSchema>;

/**
 * Baseline event-action constants covering the generic surface every app shares.
 * Emitters are NOT limited to this set — any consumer defines its own namespaced
 * actions (see {@link AuditEventInput.action}). The dashboard, alerts, and
 * out-of-the-box compliance reports pivot off these names, so keep them stable.
 */
export const AUDIT_ACTIONS = {
  // Identity
  USER_SIGNIN: "user.signin",
  USER_SIGNOUT: "user.signout",
  USER_PASSWORD_CHANGED: "user.password_changed",
  USER_INVITED: "user.invited",
  // Org
  ORG_CREATED: "org.created",
  ORG_MEMBER_ADDED: "org.member_added",
  ORG_MEMBER_REMOVED: "org.member_removed",
  ORG_ROLE_CHANGED: "org.role_changed",
  ORG_SUBSCRIPTION_CHANGED: "org.subscription_changed",
  ORG_PRODUCT_PURCHASED: "org.product_purchased",
  // Agent
  AGENT_CONFIG_CHANGED: "agent.config_changed",
  AGENT_KNOWLEDGE_DOC_ADDED: "agent.knowledge_doc_added",
  AGENT_KNOWLEDGE_DOC_REMOVED: "agent.knowledge_doc_removed",
  AGENT_ESCALATION_CREATED: "agent.escalation_created",
  AGENT_TOOL_FAILED: "agent.tool_failed",
  // CRM
  CRM_CONTACT_CREATED: "crm.contact_created",
  CRM_CONTACT_MERGED: "crm.contact_merged",
  CRM_CONTACT_DELETED: "crm.contact_deleted",
  // API auth
  API_KEY_MINTED: "api.key_minted",
  API_KEY_ROTATED: "api.key_rotated",
  API_KEY_REVOKED: "api.key_revoked",
  // Integrations
  INTEGRATION_CONNECTED: "integration.connected",
  INTEGRATION_DISCONNECTED: "integration.disconnected",
} as const;

export type AuditAction = (typeof AUDIT_ACTIONS)[keyof typeof AUDIT_ACTIONS];

/**
 * Validates the `namespace.verb` action convention: a lowercase namespace and at
 * least one lowercase verb segment, dot-separated (e.g. `crm.contact_created`,
 * `google.gmail.message_sent`). Consumers emitting their own actions should
 * assert this at their trust boundary; the canonical serialization treats
 * `action` as an opaque string, so this is a convention check, not a hard schema
 * constraint.
 */
export function isNamespacedAction(action: string): boolean {
  return /^[a-z][a-z0-9_]*(\.[a-z][a-z0-9_]+)+$/.test(action);
}
